proposals/005-filter-api-to-expose-client-and-server-tls-info.md

```diff
@@ -0,0 +1,111 @@
+
+# Authentication APIs
+
+This proposal describes a set of public APIs to expose client identity to plugin implementations.
+It makes use of [proposal-004](proposal-004) for the terminology used.
+
+
+## Current situation
+
+The `Filter` and `FilterContext` APIs currently don't directly expose any authenticated client identity information.
+Specifically if clients are authenticating using SASL, the only way a `Filter` can know about that is by intercepting those frames.
+* identity-using filters in the chain must _each_ implement SASL passthrough inspection. 
+* but this is usually incompatible with use of a filter performing SASL Termination or SASL Initiation.
+
+## Motivation
+
+The lack of API support makes implementing client identity aware plugins difficult, or impossible.
+
+Goals: 
+
+* Allow the possibility for new kinds of KRPC intercepting plugins in the future by not assuming that `Filters` are the only kind of KRPC intercepting plugin. We'll use the term **plugin**, unless saying something specifically about `Filters`.
+* Allow access to TLS-specific details by plugins should they need them.
+* Provide a flexible API to make serving niche use cases possible (though perhaps not simple).
+
+## Proposal
+
+### API for Filters to access client TLS information
+
+TLS (in contrast to SASL) is handled entirely by the proxy runtime. 
+By the time a `Filter` is instantiated the proxy already has an established TLS connection.
+All that's required is an API for exposing appropriate details to `Filters`.
+
+The following method will be added to the existing `FilterContext` interface:
+
+```java
+    /**
+     * @return The TLS context for the connection between the Kafka client and the proxy, 
+     * or empty if the client connection is not TLS.
+     */
+    Optional<ClientTlsContext> clientTlsContext();
+```
+
+Where `ClientTlsContext` is a new interface in the new package `io.kroxylicious.proxy.tls`:
+
+```
+package io.kroxylicious.proxy.tls;
+
+import java.security.cert.X509Certificate;
+import java.util.Optional;
+
+public interface ClientTlsContext {
```

Comment:
While I have fewer issues with calling this `ClientTlsContext` than I do with `ServerTlsContext` I think this is better named `DownstreamTlsContext`.